Initial support for inline types

MichaelHaas99 opened the pull request.

davleopo commented on compiler/src/jdk.graal.compiler.test/src/jdk/graal/compiler/core/test/CheckObjectEqualsTest.java:
```diff
@@ -0,0 +1,93 @@
+    public void testLeftAlwaysNull() throws InvalidInstalledCodeException {
+        resetCache();
+        for (int i = 0; i < 10000; i++) {
+            test("snippetLeftAlwaysNull", NULL, testObject1);
```
you can also call test with expected deoptimizations so not sure you need the explicit compile below with the installed code as well

MichaelHaas99 replied:
u sure? think you can only check if a deoptimization was not triggered. I used CheckCastWithNullTest.java as a template

davleopo replied:
right, only the negative has API, ok then

davleopo commented on compiler/src/jdk.graal.compiler/src/jdk/graal/compiler/core/common/type/Stamp.java:
```diff
@@ -272,7 +285,7 @@ public boolean neverDistinct(Stamp other) {
         Constant constant = this.asConstant();
         if (constant != null) {
             Constant otherConstant = other.asConstant();
-            return otherConstant != null && constant.equals(otherConstant);
```
why drop the null check here?

MichaelHaas99 replied:
Think it was automatically refactored by the IDE, I will insert the null check again

davleopo commented on compiler/src/jdk.graal.compiler/src/jdk/graal/compiler/nodes/extended/FixedInlineTypeEqualityAnchorNode.java:
```diff
@@ -0,0 +1,71 @@
+
+    @Override
+    public Node canonical(CanonicalizerTool tool) {
+        if (tool.allUsagesAvailable() && hasNoUsages()) {
```
should we also drop it if the input is known to never be a value type?

MichaelHaas99 replied:
probably better compared to doing it in the canoncial method of the ObjectEqualsNode

davleopo replied:
why? Id say in both we should do it

MichaelHaas99 replied:
okay let's do it then

MichaelHaas99 committed:
Handle null-restricted, non-flat inline type fields correctly

MichaelHaas99 committed:
Added some comments

MichaelHaas99 committed:
Declared new entry points

MichaelHaas99 committed:
Added test case for calling convention

MichaelHaas99 committed:
Adaption of test case for calling convention

MichaelHaas99 committed:
Adaptions to return inline objects scalarized in multiple registers in amd64 according to the new return convention

MichaelHaas99 committed:
First basic implementation that supports the scalarized return convention, without allowing inlining yet. Also changed PEA such that an existing inline oop, if exists, is taken instead of performing a new allocation.

MichaelHaas99 committed:
Allow inlining for the new return convention with the help of canonicalization. Some small changes. Added test cases for demonstration.

MichaelHaas99 committed:
Removed wrong last node for new return convention

MichaelHaas99 committed:
Moved changes on invoke for new return convention to plugin

MichaelHaas99 committed:
Only virtualize ReturnNode if it returns scalarized values

MichaelHaas99 committed:
Fixed bug in LIR framestate

MichaelHaas99 committed:
Changed PEA to use states with oopOrHub field set in a merge

MichaelHaas99 committed:
Remove virtual object mapping from framestate (produced by scalarized return) when frame is duplicated with pop. Fixed bug when lowering CommitAllocationOrReuseOopNode. Remove StateSplit on InlineTypeNode, doesn't carry a framestate anymore.

MichaelHaas99 committed:
Fixed wrong check for already allocated oop

MichaelHaas99 committed:
Lower CommitAllocationNodes directly

MichaelHaas99 committed:
Refactored InlineTypeNode

MichaelHaas99 committed:
Moved ProjNode into a new file

MichaelHaas99 committed:
Made ProjNode more generic

MichaelHaas99 committed:
Created a node type for scalarized returns

MichaelHaas99 committed:
Refactored lowering of CommitAllocationOrReuseOopNode

MichaelHaas99 committed:
Refactored lowering of InlineTypeNode

MichaelHaas99 committed:
Refactored scalarized invoke

MichaelHaas99 committed:
Refactored changes on PEA, added some comments

MichaelHaas99 committed:
Added comments to framestate code change during duplication

MichaelHaas99 committed:
Added comments to changes in LIRFrameState and DebugInfo

MichaelHaas99 committed:
ProjNode should be a FloatingNode

MichaelHaas99 committed:
Fixed bug isInit should be result of null check of oop

MichaelHaas99 committed:
Renamed non null information in debug info to isNotNull. Refactoring of isNotNull in InlineTypeNode. Added some comments

MichaelHaas99 committed:
Handle inlining of invokes with a scalarized return correctly (delete the InlineTypeNode and ProjNodes)

MichaelHaas99 committed:
Canonicalization on InlineTypeNode causes error in TestCallingConvention JTREG test

MichaelHaas99 committed:
Handle invoke with scalarized return directly in BytecodeParser instead of plugin

MichaelHaas99 committed:
Allow canonicalization of InlineTypeNode and ProjNode

MichaelHaas99 committed:
Canonicalization of InlineTypeNode causes error, avoid for now

MichaelHaas99 committed:
Use GraphUtil function when searching the last framestate

MichaelHaas99 committed:
Use the new InlineTypeNode also for load operations on flat fields and arrays. No more problems with side effects then.

MichaelHaas99 committed:
Use isNotNull info for a null check instead of oop when virtualizing InlineTypeNode

MichaelHaas99 committed:
Created ProjNode with additional input to keep oopOrHub node alive

MichaelHaas99 committed:
Added test126 from TestLWorld which failed in jtreg

MichaelHaas99 committed:
Added some comments and a new function to InlineTypeNode

MichaelHaas99 committed:
Remove inlining in array copy snippets for the moment

MichaelHaas99 committed:
Created method in FrameStateBuilder that expects inline type parameters to be scalarized

MichaelHaas99 committed:
Started to create helper functions to emit the new method entry points in the new calling convention

MichaelHaas99 committed:
Added possibility to set a stack increment for stack repair in frame map.
Repair stack on method leave.
Added helper function to load an oop from memory to unpack an inline object.
Produce correct code for the new entry points.

MichaelHaas99 committed:
Updated valhalla tests

MichaelHaas99 committed:
Save rootMethod in CompilationResultBuilder before code emit, will be used to know if stack repair is necessary

MichaelHaas99 committed:
Use scalarized calling convention in GraphEncoder and HostBackend

MichaelHaas99 committed:
Create a correct framestate for the StartNode for the new calling convention. Scalarized parameters are inputs to the InlineTypeNode

MichaelHaas99 committed:
Lowering an invoke node should use the correct signature. Fixed bug when lowering InlineTypeNode

MichaelHaas99 committed:
Graph should be able to save an update start position

MichaelHaas99 committed:
Fixed bugs during unpacking in entry point, handle compressed oops correctly.

MichaelHaas99 committed:
Fixed bug in unpacking for floats

MichaelHaas99 committed:
Fixed bug in unpacking wrong signature access

MichaelHaas99 committed:
Jump to inline_ro_entry on virtual call with scalarized parameters. Use an assertion when the receiver is scalarized in the bytecodeparser

MichaelHaas99 committed:
Create unverified entry point if method is non-static

MichaelHaas99 committed:
Added more test cases for new calling convention

MichaelHaas99 committed:
Set entry points to default values if no code will be produced for them

MichaelHaas99 committed:
Fixed some bugs in generation of entry points and unpacking

MichaelHaas99 committed:
Small changes tried to fix errors. Added test case

MichaelHaas99 committed:
Switched branch successors

MichaelHaas99 committed:
Use correct offsets for stack during unpacking

MichaelHaas99 committed:
Align the stack if it is extended

MichaelHaas99 committed:
Fixed wrong null check creation in InlineTypeNode

MichaelHaas99 committed:
Create a frame for the nmethod entry barrier on each entry point before unpacking

MichaelHaas99 committed:
Fixed bug wrong deoptimization condition for inlinetypenode

MichaelHaas99 committed:
Added G1ReferentFieldReadBarrier for unpacking an inline object with object fields in the entry point

MichaelHaas99 committed:
Fixed bug don't use stack repair for return on dummy frame for nmethod barrier

MichaelHaas99 committed:
Fixed wrong stack increment

MichaelHaas99 committed:
Fixed bug wrong variable as destination register

MichaelHaas99 committed:
Fix alignment of entry points, only create nmethod entry barrier for dummy frames, wrong condition for stack repair on return

MichaelHaas99 committed:
Fixed bug set stack increment also for dummy frame

MichaelHaas99 committed:
Emit move instructions after invoke with scalarized return before last instruction of the current block. Directly append isNotNull read. to invoke node instead of oopOrHub

MichaelHaas99 committed:
Avoid the scheduling problem of the InvokeWithExceptionNode with scalarized return values by excluding it from the scheduler. During the generation of the invoke, the scalarized values will also be processed before the last block instruction so this can't cause any errors anyway.

MichaelHaas99 committed:
Increase instruction distance for exception edge

MichaelHaas99 committed:
Make sure scalarized parameters are handled correctly when inlining is performed during the inlining phase. Either the target method shouldn't scalarize it's parameters or scalarization has to be done additionally before an invoke due to optimizations.

MichaelHaas99 committed:
Fix null pointer exception caused by not assigning branch probability in LogicNode during virtualization of object ObjectEqualsNode

MichaelHaas99 committed:
Check for needed scalarization in InliningUtil when invoke call target is replaced.
Remove InlineTypeNode only if the method has a scalarized return.
Adapt constant folding of LoadIndexedNode to consider flat arrays.
Set the stamp for loads on flat arrays correctly.

MichaelHaas99 committed:
Avoid receiver check when inlining with an exact inlining info and the concrete method has a scalarized receiver.

MichaelHaas99 committed:
Interface enter method including a dummy frame boolean parameter.

MichaelHaas99 committed:
Refactoring of backend changes, added some comments.

MichaelHaas99 committed:
Renamed ProjNode to ReadMultiValueNode, some refactoring.

MichaelHaas99 committed:
Fixed bug only convert type to flat array if it is known to be a flat array.

MichaelHaas99 committed:
Set max exception edge distance to reasonable number.

MichaelHaas99 committed:
Fixed bug, zero rax after invoke if it contains a klass pointer to avoid G1 OopMap error. Provide access to existing oop during PEA.

MichaelHaas99 committed:
Some additional test cases copied from the valhalla repository for debugging.

MichaelHaas99 committed:
Support nullable virtual inline objects during PEA. Fixed wrong check during lowering of InlineTypeNode.

MichaelHaas99 committed:
Move null check during if_acmp of two nullable scalarized inline objects to the beginning.

MichaelHaas99 committed:
Fix bug during unpacking of inline object in method entry point, oops were not set to zero. Set Inline Entry Point to the position of Entry Point in case they share the same entry.

MichaelHaas99 committed:
Adapt virtualization of IsNullNode and InstanceOfNode to the concept of nullable scalarized inline objects.

MichaelHaas99 committed:
PiNode should be able to change the state of virtual object to non-null during virtualization.

MichaelHaas99 committed:
Fixed bug in condition during lowering of InlineTypeNode. Duplicate framestate after traversal. Created a snippet that decides what value is returned in the first register of a scalarized return (oop, null or tagged hub).

MichaelHaas99 committed:
Use the InlineTypeNode during lowering of a commit alloc.

MichaelHaas99 committed:
Fixed bugs in substitutability check during PEA.

MichaelHaas99 committed:
Change stamp of ReturnResultDeciderNode to word stamp. Remove unnecessary function from VirtualizerTool. Rename the oop field in the ObjectState.

MichaelHaas99 committed:
Use the stamp of a virtual object to check for nullability before creating a guard during virtualization

MichaelHaas99 committed:
Fixed bug in BoxingSnippets caused by new location identity for flat arrays. Use the null check method of the GraphBuilderContext in the InlineTypePlugin.

MichaelHaas99 committed:
Do correct materialization for inline objects (create diamond). Change deoptimization action for null check necessary in null-restricted arrays. Fixed assumed bug during compute allocation order.  Simplify inline type node if constant null.

MichaelHaas99 committed:
Created InlineTypeUtil. Only scalarize parameters in root methods, to avoid erros caused by direct inlining in the BytecodeParser. Scalarize inline objects on invokes of methodHandles.

MichaelHaas99 committed:
Move utility functions into InlineTypeUtil and use them. Changed lowering method of InlineTypeNode and made sure existingOop is set in case isNotNull is also set. Make sure MethodHandle inline type parameters stay boxed if the handle was resolved.

MichaelHaas99 committed:
Do a runtime call after an invoke to a methodhandle to handle the case a scalarized inline object is returned.

MichaelHaas99 committed:
Refactoring

MichaelHaas99 committed:
Anchor reads from a nullable scalarized inline object.

MichaelHaas99 committed:
Use a better DeoptimizationReason for the ObjectEqualsSnippet.

MichaelHaas99 committed:
Deleted FixedprojAnchorNode.

MichaelHaas99 committed:
Commit before rebase.

MichaelHaas99 committed:
Fixed bugs after rebasing.
Correct wrong usage of G1ReferentFieldReadBarrier in method entry point. Dont' scalarize parameters in OSR compilation and don't do a stack repair.

MichaelHaas99 committed:
Refactoring and added comments. Implemented a provider that makes Valhalla specific options available.

MichaelHaas99 committed:
Improve object equals comparison during PEA.

MichaelHaas99 committed:
Local jtreg test changes

MichaelHaas99 committed:
Set outer declaring class for flat fields correctly to make constant folding possible.

MichaelHaas99 committed:
Added test case for checking correct constant folding on flat fields.

MichaelHaas99 committed:
Don't append foreign call if the method handle is resolved and known to not have a scalarized return.

MichaelHaas99 committed:
Provide the ValhallaEnabled option.

MichaelHaas99 committed:
Some refactoring on inline type stamp checks.

MichaelHaas99 committed:
Make unimplemented unsafe operations to be investigated.

MichaelHaas99 committed:
Avoid additional parameter in Providers constructor, to avoid problems with subclasses from other modules.

MichaelHaas99 committed:
Use default Valhalla options provider (all flags set to false) for other providers.

MichaelHaas99 committed:
Use the CommitAllocationNode if Valhalla is turned off.

MichaelHaas99 committed:
Emit an explicit IdentityException.

MichaelHaas99 committed:
Use the StampTool when checking for an inline type. InvalidateReprofile after deoptimization on monitorenter.

MichaelHaas99 committed:
Value classes should have identity if valhalla is disabled.

MichaelHaas99 committed:
Removed unused function, added assertion in ObjectEqualsNode.

MichaelHaas99 committed:
Pull-request review.

MichaelHaas99 committed:
Reevaluate substitutability check during canonicalization with valhalla options. Fixed spelling error.

MichaelHaas99 committed:
Move the stamp checks for an inline type array into StampTool.

MichaelHaas99 committed:
Additional function in StampTool which directly takes a node as parameter for checking if its stamp represents an inline type.

MichaelHaas99 committed:
Don't create a null check on the value to be stored in an array if Valhalla is disabled.

MichaelHaas99 committed:
Use the C2 compiler option to enable/disable the usage of the acmp profiling data in Graal

MichaelHaas99 committed:
Removed unused code, refactored some names and comments.

MichaelHaas99 committed:
Remove inline type methods from Stamp base class.

MichaelHaas99 committed:
Bail out in scheduler in case graph is broken.

MichaelHaas99 committed:
Some refactoring in PEA.

MichaelHaas99 committed:
Raised MAX_EXCEPTION_EDGE_OP_DISTANCE_FROM_END.

MichaelHaas99 committed:
Also delete foreign call and membar after methodhandle was resolved to a non-scalarized return triggered by simplification. Some refactoring.

MichaelHaas99 committed:
Do a runtime call for the IdentityException. Changed DeoptimizationReason.

MichaelHaas99 committed:
The base offset of flat arrays is now the same as that used for object arrays.

MichaelHaas99 committed:
Correct cycle and size of new Valhalla nodes.

MichaelHaas99 committed:
Deleted old unused nodes and snippets.

MichaelHaas99 committed:
Renamed calling convention function. Fixed typo in return convention option. Added option for calling convention.

MichaelHaas99 committed:
Use valhalla option in KlassLayoutHelperNode. Removed some comments.

MichaelHaas99 committed:
Added and adapted comments in backend.

MichaelHaas99 committed:
Removed unused code. Added and adapted comments.

MichaelHaas99 committed:
Only disallow folding of GetClassNode if valhalla is enabled.

MichaelHaas99 committed:
Fixed bug in HasIdentitySnippet and AbstractObjectStamp. Correct LocationIdentity in IsFlatArraySnippets.

MichaelHaas99 committed:
Accidentally deleted additional offset for elements in flat arrays.

MichaelHaas99 committed:
Fixed GraalInvariants errors.

MichaelHaas99 committed:
Use the old defaultPrototypeMarkWord in case Valhalla is disabled.

MichaelHaas99 committed:
Make an extra snippet for the valhalla identity hashcode.

MichaelHaas99 committed:
Fixed errors which occured during libgraal build process. Use the valhalla options first when checking on the calling and return convention. Make sure to not remove an object mapping from a framestate if it was not created for a scalarized return.

MichaelHaas99 committed:
Put statement in block. Correct wrong comment.

MichaelHaas99 committed:
Small test case adaptions used for debugging.

MichaelHaas99 committed:
Insert a piCastNonNull for hasIdentity snippet.

MichaelHaas99 committed:
Clear memory barrier during InlineTypeNode lowering.

MichaelHaas99 committed:
Added isInlineTypeOrNull utility function for object stamps.

MichaelHaas99 committed:
Canonicalize HasIdentityNode.

MichaelHaas99 committed:
Only use markword if the object is not locked.

MichaelHaas99 committed:
Also pass the stamp to the DelayedRawComparisonNode to do substitutability check recursively.

MichaelHaas99 committed:
Renamed profiling input to snippet.

MichaelHaas99 committed:
Refactored initialization or register state in backend.

MichaelHaas99 committed:
Renamed the field scalarizedInlineObject in the InlineTypeNode to fieldValues.

MichaelHaas99 committed:
Renamed existingOop to oop.

MichaelHaas99 committed:
Renamed buffered to allocated.

MichaelHaas99 committed:
Renamed null-free to non-null. Added inferStamp for InlineTypeNode.

MichaelHaas99 committed:
Renamed isNotNull to nonNull.

MichaelHaas99 committed:
Do read elimination with an anchor node for nullable virtual objects.

MichaelHaas99 committed:
Put statements in a { }, removed old comments.

MichaelHaas99 committed:
Added TODO to JTTTest.

MichaelHaas99 committed:
Adapted comments.

MichaelHaas99 committed:
Renamed ForeignCallDescriptors.

MichaelHaas99 committed:
Corrected old oopOrHub naming.

MichaelHaas99 committed:
Restored original formatting in suite.py

MichaelHaas99 committed:
Updated README.

MichaelHaas99 committed:
Some refactoring in InlineTypePlugin.

MichaelHaas99 committed:
Moved lowering of flat store nodes from DefaultJavaLoweringProvider to DefaultHotSpotLoweringProvider.

MichaelHaas99 committed:
Stop GraphUtil.unproxify at PiNode during ReadElimination, if it represents a transition from nullable to non-null stamp.

MichaelHaas99 committed:
Created a phase that replaces the original arguments of a method with the scalarized arguments according to the Valhalla Calling Convention. The scalarized arguments created during parsing are saved in the MethodCallTargetNode in the meantime.

MichaelHaas99 committed:
Make a null-check during parsing for load field operations on objects without identity to avoid it during PEA.

MichaelHaas99 committed:
Unsafe loads don't have a null-check. Remove nullCheck function from VirtualizerTool.

MichaelHaas99 committed:
Don't use a scalarized return for an inlinee graph.

MichaelHaas99 committed:
Only run ValhallaCallingConventionPhase if the calling convention is enabled.

MichaelHaas99 committed:
New method in ReturnScalarizedNode to replace a normal ReturnNode, not used at the moment.

MichaelHaas99 committed:
Update GraphState after ValhallaCallingConventionPhase. After this phase replace the arguments on the call target node directly if devirtualization happens. Only use InlineTypePlugin if valhalla is enabled.

MichaelHaas99 committed:
Some refactoring.

MichaelHaas99 committed:
Some test case changes.

MichaelHaas99 committed:
Created isUnlocked function for snippets. No need to check locking in hasIdentity check. Inserted a sanity check for array creation.

MichaelHaas99 committed:
Reevaluate substitutability check during lowering.

MichaelHaas99 committed:
Option to control the inlining of the substitutability check.

MichaelHaas99 committed:
No need to check first general register on scalarized return if the oop is not used anyway.

MichaelHaas99 committed:
Fixed SIGSEV caused in assertion due to missing entry in register map.

MichaelHaas99 committed:
Use StampTool.isInlineType during virtualization of LoadFieldNode.

MichaelHaas99 committed:
Use a PublishWriteNode when lowering an InlineTypeNode.

MichaelHaas99 committed:
IsFlatArrayNode and IsNullFreeArrayNode can be virtualized which kills CFG.

MichaelHaas99 committed:
Use StampTool.isInlineTypeArray after null-check.
Avoid materialization in merge when loading from an inline type array.

MichaelHaas99 committed:
Changed some comments.
Create the non-null check in the scalarization function.

MichaelHaas99 committed:
Virtualize from an inline object by loading its value and creating an InlineTypeNode.

MichaelHaas99 committed:
Fixed bug, type is not resolved yet.

MichaelHaas99 committed:
Virtualize materialized inline objects during PEA to avoid materializations in merges.

MichaelHaas99 committed:
Put assertion after inline type check

MichaelHaas99 committed:
Fixed bug, InlineTypeNode is not only used for Invoke and Parameter, therefore use the alias during PEA for virtualization.

MichaelHaas99 committed:
Renamed notNull to nonNull, also use the alias.

MichaelHaas99 committed:
Fixed error causing missing mapping for VirtualObjectNode in Framestate. Clear the cached state after clone when creating a new virtual object in a PiNode.
Fixed bug obsolete nodes are still referenced, use the alias for oop and nonNull when creating a virtual object.

MichaelHaas99 committed:
Refactored implementation to virtualize a materialized inline object during PEA. Use the correct effects. Combine read elimination cache and scalarization cfg.

MichaelHaas99 committed:
Resolve field in InlineTypePlugin.

MichaelHaas99 committed:
Don't use a FixedValueAnchor for a load on a virtual inline object where the entry is also virtual.

MichaelHaas99 committed:
Use caching for the additional phi nodes (nonNull and oop). Process virtual at loop exit with additional fields. Disable virtualization from inline object for now when caching is.

MichaelHaas99 committed:
Small refactoring for PEA changes.

MichaelHaas99 committed:
Correct wrong index, no materialization happens if the virtual object was already allocated once.

MichaelHaas99 committed:
ObjectState equality comparison should include oop and nonNull info.

MichaelHaas99 committed:
Correct wrong usage of effects, use effects for preprocessor block. Current implementation which virtualizes from an inline object also allows loop processing.
Avoid alloc for phi in diamond by scalarizing constant null value.

MichaelHaas99 committed:
Adapted MonitorEnterNode to nullable virtual objects.

MichaelHaas99 committed:
Renamed isInlineTypeOrNull to isNullableInlineType.

MichaelHaas99 committed:
Also virtualize inline type entries of a virtual object during object state merging.

MichaelHaas99 committed:
Fixed bug in virtualize of ReturnScalarizedNode, make sure the field values are materialized and are not replaced with a virtual object.

MichaelHaas99 committed:
Toogle condition.

MichaelHaas99 committed:
Use the alias after ensureMaterialize.

MichaelHaas99 committed:
Missing assignment of state entry with merged virtual inline object.

MichaelHaas99 committed:
Fixed bug, wrong stamp information for oop and nonNull phis.

MichaelHaas99 committed:
Fixed bug, virtual object node was passed as field entry to read cache.

MichaelHaas99 committed:
Added simplification for CommitAllocationOrReuseOopNode.

MichaelHaas99 committed:
Small modifications on local Valhalla test cases to try out stuff.

MichaelHaas99 committed:
Compute a correct stamp for the merged oop.

MichaelHaas99 committed:
Added isIdentity and isValue functions from Class.java to DynamicHub.java, otherwise substitution would be missing.

MichaelHaas99 committed:
Hotspot backend already enters a frame during code prefix emission. Don't enter a frame during LIR emission only for Hotspot.

MichaelHaas99 committed:
Moved all hash code logic for value objects into the HotSpotHashCodeSnippets.

MichaelHaas99 committed:
Fixed wrong usage of PublishWritesNode in combination with inline objects and diamonds.

MichaelHaas99 committed:
An inline type array can't have an abstract class or interface as component type.

MichaelHaas99 committed:
Also consider SnippetObjectConstant in delayed raw comparison.

MichaelHaas99 committed:
Use an empty array if the substitutability check is not inlined.

MichaelHaas99 committed:
Deoptimize if a nullable flat field is recognized.

MichaelHaas99 committed:
Avoid missing types error for ObjectEqualsSnippets in libgraal.

MichaelHaas99 committed:
Fixed bugs caused by virtualize from inline object implementation. Too many iterations at loop due to non-cached virtual objects created for entries during object state merge.  Request on cached phis should be avoided in recursive call.

MichaelHaas99 committed:
Fixed bugs caused by adaptions in HotSpotHashCodeSnippets.

MichaelHaas99 committed:
Add types needed by ObjectEqualsSnippet in a correct way.

MichaelHaas99 committed:
Adapt caching for virtualize from inline object.

MichaelHaas99 committed:
Some adaptions on virtualize from inline object implementation. Only use phi cache for inline types. Don't perform it in state merge when entries have different types.

MichaelHaas99 committed:
Don't constant fold the hash code of hotspot objects if we are in libgraal. They are computed by the object handle and only valid for internal usage.

MichaelHaas99 committed:
The ObjectEqualsSnippet  compares two tracked pointers. Tautology if both operands refer to the same node.

MichaelHaas99 committed:
Do a null-check for stores to a null-restricted inline type field.

MichaelHaas99 committed:
Made valhalla array layout check nodes floating (similar to ObjectIsArrayNode with an intrinsic similar to InstanceofNode) to allow value numbering.

MichaelHaas99 committed:
Updated flat write nodes to have multiple memory kills.
Updated virtualize of StoreFlatFieldNode to use alias for value.

MichaelHaas99 committed:
Renamed StoreFlatIndexNode to StoreFlatElementNode.

MichaelHaas99 committed:
Value class foreign calls are safepoints, also corrected LocationIdentity. Insert a membar after load from unknown inline type array.

MichaelHaas99 committed:
Created isAllocatedOrNull field for InlineTypeNode.

MichaelHaas99 committed:
Avoid adding an already allocated or null oop to the graph during PEA.

MichaelHaas99 committed:
Avoid virtualization in ArrayCopyNode for inline type arrays.

MichaelHaas99 committed:
Use curly brackets around statements.

MichaelHaas99 committed:
Most value class foreign calls do not kill a location.

MichaelHaas99 committed:
Remove "Node" suffix in name template.

MichaelHaas99 committed:
Anchor LoadIndexed nodes, otherwise they float above flat array check.

MichaelHaas99 committed:
Local test case updates.

MichaelHaas99 commented:
This PR is outdated.
Link to current PR https://github.com/MichaelHaas99/graal/pull/2
